### DISCUSSION TITLE
I am interested in developing bindings of this crate for c++(mainly) and python3. Any pointers on where should I start.?

### DISCUSSION THREAD

**Discussion by saurabh10041998** [bgpkit/bgpkit-parser#179] [Q&A] [answered]
Hii All,
I am impressed by the range of BGP message types and methods supported by the crate.
Also thanks for actively maintaining the project.

I am interested in developing bindings of this crate for c++ (mainly) and python3. Can anyone provide me pointers where should i start ? and how should I approach this ?

Thanks

**Comment by digizeph** [comment, marked as answer]
Hey @saurabh10041998, thank you for your kinds words!

For Python3 binding, I have been maintaining [`bgpkit-parser-py`](https://github.com/bgpkit/bgpkit-parser-py), and [`pybgpkit`](https://github.com/bgpkit/pybgpkit) that uses it. I used [`pyo3`](https://github.com/PyO3/pyo3) library to build the bindings. Bindings do not cover everything the Rust library offers, and only targeted for basic usages (loop through messages and access bgp msg content). You can try it out and I'd be happy to take questions or PRs on these projects. Helps with documentation and examples would also be highly appreciated.

I am not sure about C++ though as I have no practical experience with it. I'd expect it would be similar to calling Rust from C (https://doc.rust-lang.org/nomicon/ffi.html#calling-rust-code-from-c). Maybe give it a try and see if you can get started there.

Cheers,
Mingwei
